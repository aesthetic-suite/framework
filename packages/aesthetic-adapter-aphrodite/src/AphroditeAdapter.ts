/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import {
  Adapter,
  ClassName,
  Declaration,
  UnifiedSyntax,
  injectKeyframes,
  injectFontFaces,
  StyleSheetMap,
} from 'aesthetic';
import { StyleSheet as Aphrodite, Extension } from 'aphrodite';
import { NativeDeclaration as Properties, ParsedDeclaration } from './types';

export default class AphroditeAdapter extends Adapter<ParsedDeclaration> {
  aphrodite: {
    css(...styles: ParsedDeclaration[]): ClassName;
    StyleSheet: typeof Aphrodite;
  };

  constructor(extensions: Extension[] = []) {
    super();

    this.aphrodite = Aphrodite.extend([
      ...extensions,
      { selectorHandler: this.handleHierarchySelector },
      { selectorHandler: this.handleGlobalSelector },
    ]);
  }

  bootstrap(syntax: UnifiedSyntax<Properties>) {
    syntax.on('property', this.handleProperty); //.on('@global', this.handleGlobal);
  }

  createStyleSheet(styleSheet: any): StyleSheetMap<ParsedDeclaration> {
    return this.aphrodite.StyleSheet.create(styleSheet) as StyleSheetMap<ParsedDeclaration>;
  }

  transformToClassName(...styles: ParsedDeclaration[]): ClassName {
    const legitStyles: ParsedDeclaration[] = [];
    const tempStylesheet: { [key: string]: any } = {};
    let counter = 0;

    styles.forEach(style => {
      // eslint-disable-next-line no-underscore-dangle
      if (style._name && style._definition) {
        legitStyles.push(style);
      } else {
        tempStylesheet[`inline${counter}`] = style;
        counter += 1;
      }
    });

    if (counter > 0) {
      legitStyles.push(...Object.values(this.create(tempStylesheet)));
    }

    return this.aphrodite.css(...legitStyles);
  }

  // handleGlobal(styleSheet: StyleSheet, declaration: Declaration, selector: string) {
  //   styleSheet.globals = {
  //     ...styleSheet.globals,
  //     [`*${selector}`]: declaration,
  //   };
  // }

  handleGlobalSelector(
    selector: string,
    baseSelector: string,
    callback: (selector: string) => string,
  ): string | null {
    if (selector.charAt(0) !== '*') {
      return null;
    }

    return callback(selector.slice(1));
  }

  handleHierarchySelector(
    selector: string,
    baseSelector: string,
    callback: (selector: string) => string,
  ): string | null {
    if (selector.charAt(0) !== '>' && selector.charAt(0) !== '[') {
      return null;
    }

    return callback(`${baseSelector}${selector}`);
  }

  handleProperty(declaration: Declaration<Properties>, property: string, value: any) {
    if (property === 'animationName') {
      declaration[property] = injectKeyframes(value, this.unifiedSyntax!.keyframes);
    } else if (property === 'fontFamily') {
      declaration[property] = injectFontFaces(value, this.unifiedSyntax!.fontFaces);
    } else {
      declaration.addProperty(property, value);
    }
  }
}
