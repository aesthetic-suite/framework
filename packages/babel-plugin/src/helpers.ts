import { types as t } from '@babel/core';
import { createDebugger } from '@boost/debug';

export const debug = createDebugger('aesthetic:babel');

export function isFunction(node: t.Node) {
  return t.isFunctionExpression(node) || t.isArrowFunctionExpression(node);
}
