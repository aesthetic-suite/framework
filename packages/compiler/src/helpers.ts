import { ColorShade } from '@aesthetic/system';

export function formatShade(value: number | string): ColorShade {
	return String(value).padStart(2, '0').slice(0, 2) as ColorShade;
}

export function formatUnit(value: number): string {
	return value.toFixed(2).replace('.00', '');
}
