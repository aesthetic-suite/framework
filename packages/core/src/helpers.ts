import { Theme } from '@aesthetic/system';
import { objectLoop } from '@aesthetic/utils';
import { SheetParams } from './types';

export function createCacheKey(params: SheetParams, type: string): string {
  let key = type;

  // Since all other values are scalars, we can just join the values.
  // This is 3x faster than JSON.stringify(), and 1.5x faster than Object.values()!
  objectLoop(params, (value) => {
    key += value;
  });

  return key;
}

export function createDefaultParams(theme: Theme<{}>, params: SheetParams): Required<SheetParams> {
  return {
    contrast: theme.contrast,
    deterministic: false,
    direction: 'ltr',
    scheme: theme.scheme,
    theme: theme.name,
    unit: 'px',
    vendor: false,
    ...params,
  };
}
