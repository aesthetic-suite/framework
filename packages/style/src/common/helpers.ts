/* eslint-disable no-magic-numbers */

import { Value, MaybeValue, CSS, Sheet } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';
import { IMPORT_RULE, STYLE_RULE } from './constants';

export function insertRule(sheet: Sheet, rule: CSS, index?: number): number {
  try {
    return sheet.insertRule(rule, index ?? sheet.cssRules.length);
  } catch {
    // Vendor prefixed properties, pseudos, etc, that are inserted
    // into different vendors will trigger a failure. For example,
    // `-moz` or `-ms` being inserted into WebKit.
    // There's no easy way around this, so let's just ignore the
    // error so that subsequent styles are inserted.
    // istanbul ignore next
    return -1;
  }
}

export function insertAtRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // At-rules must be inserted before normal style rules.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type === STYLE_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function insertImportRule(sheet: Sheet, rule: CSS): number {
  const { length } = sheet.cssRules;
  let index = 0;

  // Import rules must be inserted at the top of the style sheet,
  // but we also want to persist the existing order.
  for (let i = 0; i <= length; i += 1) {
    index = i;

    if (sheet.cssRules[i]?.type !== IMPORT_RULE) {
      break;
    }
  }

  return insertRule(sheet, rule, index);
}

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import';
}

export function isMediaRule(value: string): boolean {
  return value.slice(0, 6) === '@media';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

export function isSupportsRule(value: string): boolean {
  return value.slice(0, 9) === '@supports';
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      // eslint-disable-next-line no-console
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}

export function joinQueries(prev: string | undefined, next: string): string {
  if (prev) {
    return `${prev} and ${next}`;
  }

  return next;
}
