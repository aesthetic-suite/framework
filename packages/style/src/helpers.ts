/* eslint-disable no-magic-numbers */

import { Value, MaybeValue } from '@aesthetic/types';
import { hyphenate } from '@aesthetic/utils';

export function isAtRule(value: string): boolean {
  return value[0] === '@';
}

export function isImportRule(value: string): boolean {
  return value.slice(0, 7) === '@import' && value[7] === ' ';
}

export function isNestedSelector(value: string): boolean {
  const char = value[0];

  return (
    char === ':' ||
    char === '[' ||
    char === '>' ||
    char === '~' ||
    char === '+' ||
    char === '*' ||
    char === '|'
  );
}

const unitlessProperties = new Set<string>();

[
  'animationIterationCount',
  'borderImage',
  'borderImageOutset',
  'borderImageSlice',
  'borderImageWidth',
  'columnCount',
  'columns',
  'flex',
  'flexGrow',
  'flexPositive',
  'flexShrink',
  'flexNegative',
  'flexOrder',
  'fontWeight',
  'gridArea',
  'gridRow',
  'gridRowEnd',
  'gridRowSpan',
  'gridRowStart',
  'gridColumn',
  'gridColumnEnd',
  'gridColumnSpan',
  'gridColumnStart',
  'lineClamp',
  'lineHeight',
  'maskBorder',
  'maskBorderOutset',
  'maskBorderSlice',
  'maskBorderWidth',
  'opacity',
  'order',
  'orphans',
  'tabSize',
  'widows',
  'zIndex',
  'zoom',
  // SVG
  'fillOpacity',
  'floodOpacity',
  'stopOpacity',
  'strokeDasharray',
  'strokeDashoffset',
  'strokeMiterlimit',
  'strokeOpacity',
  'strokeWidth',
].forEach((property) => {
  unitlessProperties.add(property);
  unitlessProperties.add(hyphenate(property));
});

export function isUnitlessProperty(property: string): boolean {
  return unitlessProperties.has(property);
}

export function isValidValue(property: string, value: MaybeValue | boolean): value is Value {
  if (value === null || value === undefined || value === true || value === false || value === '') {
    if (__DEV__) {
      console.warn(`Invalid value "${value}" for "${property}".`);
    }

    return false;
  }

  return true;
}

export function isVariable(value: string): boolean {
  return value.slice(0, 2) === '--';
}
