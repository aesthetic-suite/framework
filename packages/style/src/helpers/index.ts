import createAtomicCacheKey from './createAtomicCacheKey';
import formatConditions from './formatConditions';
import formatDeclaration from './formatDeclaration';
import formatDeclarationBlock from './formatDeclarationBlock';
import formatRule from './formatRule';
import formatVariable from './formatVariable';
import formatVariableName from './formatVariableName';
import getDocumentStyleSheet from './getDocumentStyleSheet';
import isAtRule from './isAtRule';
import isFontFaceFule from './isFontFaceFule';
import isImportRule from './isImportRule';
import isInvalidValue from './isInvalidValue';
import isKeyframesRule from './isKeyframesRule';
import isMediaRule from './isMediaRule';
import isNestedSelector from './isNestedSelector';
import isSupportsRule from './isSupportsRule';
import isUnitlessProperty from './isUnitlessProperty';
import isVariable from './isVariable';

export {
  createAtomicCacheKey,
  formatConditions,
  formatDeclaration,
  formatDeclarationBlock,
  formatRule,
  formatVariable,
  formatVariableName,
  getDocumentStyleSheet,
  isAtRule,
  isFontFaceFule,
  isImportRule,
  isInvalidValue,
  isKeyframesRule,
  isMediaRule,
  isNestedSelector,
  isSupportsRule,
  isUnitlessProperty,
  isVariable,
};
