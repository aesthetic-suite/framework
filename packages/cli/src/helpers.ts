import { CONFIG_FOLDER } from '@aesthetic/compiler';
import { Path } from '@boost/common';
import { CWD } from './constants';

export function getConfigFolderDir(name: string): Path {
  return new Path(CWD, CONFIG_FOLDER, name);
}

export function validateSystemName(value: string) {
  if (value && !value.match(/^[-a-z0-9]+$/giu)) {
    throw new Error('Name may only contain numbers, letters, and dashes.');
  }
}
