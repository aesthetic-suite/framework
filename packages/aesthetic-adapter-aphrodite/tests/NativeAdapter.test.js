/* eslint-disable sort-keys */

import { StyleSheet, StyleSheetTestUtils } from 'aphrodite';
import AphroditeAdapter from '../src/NativeAdapter';
import {
  FONT_ROBOTO_FLAT_SRC,
  FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
  KEYFRAME_FADE,
  SYNTAX_NATIVE_PARTIAL,
  SYNTAX_ATTRIBUTE,
  SYNTAX_DESCENDANT,
  SYNTAX_PSEUDO,
} from '../../../tests/mocks';
import { renderAphroditeStyles } from '../../../tests/helpers';

describe('aesthetic-adapter-aphrodite/NativeAdapter', () => {
  let instance;

  beforeEach(() => {
    StyleSheetTestUtils.suppressStyleInjection();
    instance = new AphroditeAdapter();
  });

  afterEach(() => {
    StyleSheetTestUtils.clearBufferAndResumeStyleInjection();
  });

  it('can customize the aphrodite instance through the constructor', () => {
    const extension = { selectorHandler() {} };
    instance = new AphroditeAdapter([extension]);

    expect(instance.aphrodite).not.toEqual(StyleSheet);
  });

  it('can transform dynamic styles', () => {
    expect(
      instance.transform({
        width: 10,
        height: 10,
      }),
    ).toBe('inline0_7in6ye');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('transforms style declarations into class names', () => {
    expect(instance.transform(instance.create(SYNTAX_NATIVE_PARTIAL).button)).toBe(
      'button_13l44zh',
    );
  });

  it('combines different style declarations into unique class names', () => {
    const sheet = instance.create({
      foo: {
        color: 'red',
        display: 'block',
      },
      bar: {
        color: 'green',
        margin: 5,
      },
      baz: {
        color: 'blue',
        padding: 5,
      },
    });

    expect(instance.transform(sheet.foo)).toBe('foo_1u9pmmq');
    expect(instance.transform(sheet.bar)).toBe('bar_1etchdu');
    expect(instance.transform(sheet.baz)).toBe('baz_xw1a2w');
    expect(instance.transform(sheet.foo, sheet.baz)).toBe('foo_1u9pmmq-o_O-baz_xw1a2w');
    expect(instance.transform(sheet.bar, sheet.foo)).toBe('bar_1etchdu-o_O-foo_1u9pmmq');
  });

  it('handles globals', () => {
    instance.transform(
      instance.create({
        globals: {
          '*body': { margin: 0 },
          '*html': { height: '100%' },
          '*a': {
            color: 'red',
            ':hover': {
              color: 'darkred',
            },
          },
        },
      }).globals,
    );

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles attribute selectors', () => {
    expect(instance.transform(instance.create(SYNTAX_ATTRIBUTE).attr)).toBe('attr_424cv8');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles descendant selectors', () => {
    expect(instance.transform(instance.create(SYNTAX_DESCENDANT).list)).toBe('list_1lo5lhe');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles pseudo selectors', () => {
    expect(instance.transform(instance.create(SYNTAX_PSEUDO).pseudo)).toBe('pseudo_q2zd6k');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles font faces', () => {
    expect(
      instance.transform(
        instance.create({
          font: {
            fontFamily: [FONT_ROBOTO_FLAT_SRC],
            fontSize: 20,
          },
        }).font,
      ),
    ).toBe('font_uk6a9p');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles multiple font faces', () => {
    expect(
      instance.transform(
        instance.create({
          font: {
            fontFamily: FONT_CIRCULAR_MULTIPLE_FLAT_SRC,
            fontSize: 20,
          },
        }).font,
      ),
    ).toBe('font_f7iz6d');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles animations', () => {
    expect(
      instance.transform(
        instance.create({
          animation: {
            animationDuration: '3s, 1200ms',
            animationIterationCount: 'infinite',
            animationName: KEYFRAME_FADE,
          },
        }).animation,
      ),
    ).toBe('animation_mab5hn');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });

  it('handles media queries', () => {
    expect(
      instance.transform(
        instance.create({
          media: {
            color: 'red',
            '@media (min-width: 300px)': {
              color: 'blue',
            },
            '@media (max-width: 1000px)': {
              color: 'green',
            },
          },
        }).media,
      ),
    ).toBe('media_1yqe7pa');

    expect(renderAphroditeStyles(instance)).toMatchSnapshot();
  });
});
