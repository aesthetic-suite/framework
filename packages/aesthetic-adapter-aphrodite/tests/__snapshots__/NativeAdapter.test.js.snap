// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aesthetic-adapter-aphrodite/NativeAdapter handles animations 1`] = `"@keyframes keyframe_18jn58a{from{opacity:0;}to{opacity:1;}}.animation_mab5hn{-webkit-animation-duration:3s, 1200ms !important;animation-duration:3s, 1200ms !important;-webkit-animation-iteration-count:infinite !important;animation-iteration-count:infinite !important;-webkit-animation-name:keyframe_18jn58a !important;animation-name:keyframe_18jn58a !important;}"`;

exports[`aesthetic-adapter-aphrodite/NativeAdapter handles font faces 1`] = `"@font-face{font-family:Roboto;font-style:normal;font-weight:normal;src:local('Robo'), url('fonts/Roboto.woff2') format('woff2'), url('fonts/Roboto.ttf') format('truetype');}.font_uk6a9p{font-family:\\"Roboto\\" !important;font-size:20px !important;}"`;

exports[`aesthetic-adapter-aphrodite/NativeAdapter handles globals 1`] = `"body{margin:0px !important;}html{height:100% !important;}a{color:red !important;}a:hover{color:darkred !important;}"`;

exports[`aesthetic-adapter-aphrodite/NativeAdapter handles media queries 1`] = `".media_1yqe7pa{color:red !important;}@media (min-width: 300px){.media_1yqe7pa{color:blue !important;}}@media (max-width: 1000px){.media_1yqe7pa{color:green !important;}}"`;

exports[`aesthetic-adapter-aphrodite/NativeAdapter handles pseudos 1`] = `".pseudo_q2zd6k{position:fixed !important;}.pseudo_q2zd6k:hover{position:static !important;}.pseudo_q2zd6k::before{position:absolute !important;}"`;
